crates/subspace-farmer/src/utils/farmer_piece_getter.rs

```diff
@@ -68,6 +74,19 @@ where
             return Ok(maybe_piece);
         }
 
+        if let Some(readers_and_pieces) = self.weak_readers_and_pieces.upgrade() {
```

Comment:
Pls explain why `Weak` is needed here. Also,  the Option<> to the future in L81 lives after the lock is dropped, not sure how this works. Does this assume that once the `Option<>` inside the Mutex goes from `None` -> `Some()`, it won't go back to None?

Overall, think it would be cleaner to share an `Arc<Foo>`. where `Foo` manages a state machine and is interior mutable

Reply:
In this case it was indeed unnecessary, replaced with `Arc`.

Reply:
The second question still applies :-

Reply:
Technically it doesn't matter here. Practically yes, it only goes one way, it is necessary to break a cycle where need `PiecesAndRecords` to use in networking `Node`, but networking `Node` itself is a requirement to create `SingleDiskPlot` that makes `PiecesAndRecords` actually work.

I found a better pattern when designing new piece cache that I'll eventually use for plotted pieces as well, but it'll be a separate refactoring.

Reply:
I see ok. This is similar to the NetworkWrapper issue in block relay IIUC